app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java

```diff
@@ -0,0 +1,574 @@
+
+import android.content.ContentValues;
+import android.database.Cursor;
+import android.util.Pair;
+
+import net.sqlcipher.database.SQLiteDatabase;
+
+import org.commcare.android.crypt.EncryptionIO;
+import org.commcare.android.logic.GlobalConstants;
+import org.commcare.android.util.FileUtil;
+import org.commcare.android.util.SessionUnavailableException;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.commcare.modern.database.DatabaseHelper;
+import org.javarosa.core.io.StreamsUtil;
+import org.javarosa.core.services.storage.EntityFilter;
+import org.javarosa.core.services.storage.IStorageIterator;
+import org.javarosa.core.services.storage.Persistable;
+import org.javarosa.core.util.InvalidIndexException;
+import org.javarosa.core.util.externalizable.Externalizable;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Vector;
+
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * Sql logic for storing persistable objects. Uses the filesystem to store
+ * persitables in _encrypted_ manner; useful when objects are larger than the
+ * 1mb sql row limit.
+ *
+ * @author Phillip Mates (pmates@dimagi.com).
+ */
+public class HybridFileBackedSqlStorage<T extends Persistable> extends SqlStorage<T> {
+    private final File dbDir;
+    private final static int ONE_MB_DB_SIZE_LIMIT = 1000000;
+
+    /**
+     * Column selection used for reading file data:
+     * - Id column needed to correctly set the id of objects read from db, which isn't set at write time for efficiency.
+     * - Data column holds serialized objects under 1mb
+     * - File column points to file holding serialized object over 1mb
+     * - Aes column holds encryption key for objects saved to filesystem
+     *
+     * Constraint: we never expect both data and file/aes columns to contain data at the same time
+     */
+    protected final static String[] dataColumns =
+            {DatabaseHelper.ID_COL, DatabaseHelper.DATA_COL,
+                    DatabaseHelper.FILE_COL, DatabaseHelper.AES_COL};
+
+    /**
+     * Sql object storage layer that stores serialized objects on the filesystem.
+     *
+     * @param tableName     name of database table
+     * @param classType     type of object being stored in this database
+     * @param directoryName Name of storage root subdir where entry files are placed
+     */
+    public HybridFileBackedSqlStorage(String tableName,
+                                      Class<? extends T> classType,
+                                      AndroidDbHelper dbHelper,
+                                      String directoryName) {
+        super(tableName, classType, dbHelper);
+
+        final String subPath = GlobalConstants.FILE_CC_DB + directoryName + "/_" + tableName;
```

Comment:
Prefix `tableName` part of the directory with `_` to prevent the issue where if `tableName = ""` and we call `removeAll` it removes subdirectories associated with other tables